Rework accepting custom responses (#1746)

kowalski-se opened the pull request:
Sorry to bring this dogshit of #1746 up again, but after a while of testing I discovered that the previous approach wasn't quite good enough after all. We also have scenarios where we have a valid positive response, but the error case of the same command brings up an ill-formed response. In addition, there are multiline success responses while the error response is single-line, well, all of the possible ill-formed responses are single-liners (as far as I can tell currently).

As a result, I figured to change this `linesExpected` approach a bit to better fit these requirements. Feel free to have a look, I'd be also open to keep the special case code in my own library, in which case it would be fine to make the `GetReply` method(s) virtual so that I can inherit a subclass, overwriting the essential parts 😄

There is one small change in the handling of multiline responses (independent of the actual changes) though, previously whenever there was a "well-formed" FTP response within the multilines, the FtpReply would be handled as completed single line response. This is no longer the case, now we actually wait for the final closing line. This should technically be an improvement as per RFC, but on the other hand if we occur a disconnect in the middle of a multiline response, we're kinda stuck in the loop until timeout. Which ever way of handling this you find preferable, I'd be fine with, this is just an fyi. Example of this case could be something like this:
```
123-First line  
Second line  
  234 A line beginning with numbers  
123 The last line
```

FanDjango commented:
>  I discovered that the previous approach wasn't quite good enough after all

> As a result, I figured to change this linesExpected approach a bit to better fit these requirements

I can't say that I am surprised. I initially actually wanted to code a very lenient and flexible approach, allowing you to define all sorts of things to "accept", perhaps using some formal notation.

I'll go over your changes. A first quick look is positive. Have you run them through the integration tests?

FanDjango commented:
> even with help of the documentation

'twould be interesting where the problem lies, but.... before that:

The TimeoutTests ( ConnectTimeout ) fails. That is not an integration test. Here is the screenshot:

![image](https://github.com/user-attachments/assets/54481fef-084c-4e63-a85d-b04131fffe82)

I will revert the merge until this is cleared up.

kowalski-se commented:
Problem with the integration tests could be anything. I have a complicated setup at work, a VM on a VSphere server with couple of proxy and firewall settings. Running the `.bat` file also throws some errors like this:
```
C:\Projects\FluentFTP\FluentFTP.Dockers>docker build common-mirror --pull --progress=plain -t common-mirror:fluentftp
#0 building with "desktop-linux" instance using docker driver

#1 [internal] load build definition from Dockerfile
#1 transferring dockerfile: 348B 0.0s done
#1 DONE 0.1s

#2 [internal] load metadata for docker.io/library/python:3.10-slim
#2 ERROR: failed to do request: Head "https://registry-1.docker.io/v2/library/python/manifests/3.10-slim": EOF
------
```
Didn't have too much time to dig into every issue I occur, at this point I simply gave up.

And regarding the unit tests, as I said, all succeed on my machine after I fixed the one with the date time parsing, so no idea what to do about this 😮

![image](https://github.com/user-attachments/assets/0365207e-c9ba-4c61-8892-8b86fdfd5923)

Or if you prefer the command line execution:

```
C:\Projects\FluentFTP> dotnet test --filter Unit
Restore complete (1,8s)
  FluentFTP net5.0 succeeded (0,5s) → FluentFTP\bin\Debug\net5.0\FluentFTP.dll
  FluentFTP net472 succeeded (0,3s) → FluentFTP\bin\Debug\net472\FluentFTP.dll
  FluentFTP net462 succeeded (0,3s) → FluentFTP\bin\Debug\net462\FluentFTP.dll
  FluentFTP netstandard2.1 succeeded (0,3s) → FluentFTP\bin\Debug\netstandard2.1\FluentFTP.dll
  FluentFTP netstandard2.0 succeeded (0,4s) → FluentFTP\bin\Debug\netstandard2.0\FluentFTP.dll
  FluentFTP net6.0 succeeded (0,5s) → FluentFTP\bin\Debug\net6.0\FluentFTP.dll
  FluentFTP net7.0 succeeded (4,9s) → FluentFTP\bin\Debug\net7.0\FluentFTP.dll
  FluentFTP.Xunit succeeded (1,3s) → FluentFTP.Xunit\bin\Debug\net7.0\FluentFTP.Xunit.dll
[xUnit.net 00:00:00.00] xUnit.net VSTest Adapter v3.0.2+dd36e86129 (64-bit .NET 7.0.20)
[xUnit.net 00:00:00.59]   Discovering: FluentFTP.Xunit
[xUnit.net 00:00:00.61]   Discovered:  FluentFTP.Xunit
  FluentFTP.Xunit test succeeded with 1 warning(s) (2,0s)
    C:\Program Files\dotnet\sdk\9.0.301\Microsoft.TestPlatform.targets(48,5): warning : No test matches the given testcase filter `Unit` in C:\Projects\FluentFTP\FluentFTP.Xunit\bin\Debug\net7.0\FluentFTP.Xunit.dll
  FluentFTP.Tests succeeded (0,9s) → FluentFTP.Tests\bin\Debug\net7.0\FluentFTP.Tests.dll
[xUnit.net 00:00:00.00] xUnit.net VSTest Adapter v3.0.2+dd36e86129 (64-bit .NET 7.0.20)
[xUnit.net 00:00:00.36]   Discovering: FluentFTP.Tests
[xUnit.net 00:00:00.42]   Discovered:  FluentFTP.Tests
[xUnit.net 00:00:00.44]   Starting:    FluentFTP.Tests
[xUnit.net 00:00:06.61]   Finished:    FluentFTP.Tests
  FluentFTP.Tests test succeeded (7,6s)

Test summary: total: 45, failed: 0, succeeded: 45, skipped: 0, duration: 10,4s
Build succeeded with 1 warning(s) in 19,0s
```

FanDjango commented:
> #1 [internal] load build definition from Dockerfile
#1 transferring dockerfile: 348B 0.0s done
#1 DONE 0.1s

#2 [internal] load metadata for docker.io/library/python:3.10-slim
#2 ERROR: failed to do request: Head "https://registry-1.docker.io/v2/library/python/manifests/3.10-slim": EOF

I'll look into that as time permits

FanDjango commented:
> I assume your unit test issue resolved itself, since you haven't reverted the changes yet? 

No it did not. So I reverted, you can see that in the master's commits.

I was hoping for you to find out which part of you changes caused that timeout test to fail.

What I did do, is to keep you other change to the parser tests, where the date was used incorrectly.

> Just one more question, do you have a new NuGet release already planned?

No, I don't know about the release schedule at this moment. But that also gives you a chance to investigate the above mentioned problem, so you can get it in **before** the next release.

kowalski-se commented:
> I don't understand your reaction.

I apologize if anything came across hostile here, please note that this was not my intention.

> You have in effect removed Noop logic. I suppose you do not use NOOP. I certainly need to look into this.

I didn't. The same logic is now a condition in the while loop, see `while (isMultiline || exhaustNoop);`. I was able to do so, since there are no more operations in the loop after `DecodeStringToReply`. So I just moved the conditional continue into the exit condition.

(`while (true)` loops, as well as `break` and `continue` operations are generally considered bad practice, since they confuse the control flow).

> Listen: This thing is useful but it needs to be integrated into the overall code as all users use it, not just you.

Exactly that's the thing I'm not sure about. I want for any FOSS to be as clean, efficient and compliant as possible. While covering non-standard, even non-compliant behavior like this issue, should in my opinion not have any rights to be part of FOSS code. But perhaps this is where I'm lacking experience.

> but what is wrong with taking this as far as possible and ironing out all problems.

If you really must know, I currently just don't have any time to do that, even though I really want to. At least until Oct/Nov this year. After that I'll be happy to check up on this again, I just honestly can't at the moment.

> Oh, I reworked your rework a little in a test branch and it passes the test now.

I'm kind of interested what fixed it :)

FanDjango commented:
@kowalski-se After some organisational communication with the owner, you might consider **re-opening** your PR, which I would recommend.
